matrix/src/dense.rs

```diff
@@ -456,18 +456,55 @@ impl<T: Clone + Default + Send + Sync> DenseMatrix<T> {
     }
 }
 
-impl<T: Copy + Default + Send + Sync> DenseMatrix<T> {
-    pub fn transpose(&self) -> Self {
+impl<T: Copy + Default + Send + Sync, V: DenseStorage<T>> DenseMatrix<T, V> {
+    /// Take the transpose the matrix, returning a new matrix with the rows and columns swapped.
+    pub fn transpose(&self) -> RowMajorMatrix<T> {
         let nelts = self.height() * self.width();
         let mut values = vec![T::default(); nelts];
-        transpose::transpose(&self.values, &mut values, self.width(), self.height());
-        Self::new(values, self.height())
+        transpose::transpose(
+            self.values.borrow(),
+            &mut values,
+            self.width(),
+            self.height(),
+        );
+        RowMajorMatrix::new(values, self.height())
     }
 
-    pub fn transpose_into(&self, other: &mut Self) {
+    /// Transpose the matrix out of place copying the results into the given matrix.
```

Comment:
Nit: Pretty sure `transpose::transpose(..)` doesn't copy, it builds the result directly in `other` avoiding allocation.
```suggestion
    /// Transpose the matrix returning the result in `other` without intermediate allocation.
```